### ISSUE TITLE
Consider using a less restrictive license?

### ISSUE THREAD

**Issue by OliverGavin** [HeroicKatora/xdpilone#6]
Hello,

I was considering using this library at work but as such I cannot even test/evaluate it due to the EUPL-1.2 license. I was just curious if you would consider a less restrictive license like Apache or MIT or BSD? Technically my usage is within the license remit but it is awkward to get approval for usage.

I came across this crate in aya examples but just found it unfortunate that the licensing was more restrictive. Other teams around me may be interested in using it.

Also I completely understand if you want to stick with this license also :) 

Many thanks!

**Comment by OliverGavin** [comment]
> Are you talking about a procedural issue where EUPL primarily has not been evaluated to an allow-list within your company

This ^ (sort of)

It is not allow listed across the board because it has clauses that require sharing source code in certain cases (which don't even apply in the case of my usage because there is no distribution involved) but because of this it is not permitted for use without going through an approval process which is likely to make me seek alternatives first. Even with approval it becomes difficult to use it and nobody else can use it without seeking the same approval. I do get that that is a me problem :) but I would think other organizations handle it similarly also.


Apologies for my ignorance, I am not very well versed on various licenses and such.

**Comment by HeroicKatora** [comment]
Thanks for explaining. However, the precise extent of the sharing clause was part of the reason to choose the license. I'm willing to offer an organizational license, covering costs involved with re-licensing all contributions into such terms. Contact me as listed here: <https://www.hardmo.de/about> to discuss terms and licensing fees if you want to.

**Comment by djahandarie** [comment]
@HeroicKatora I'm not sure if you saw already, but this PR on `aya` has been in limbo for months: https://github.com/aya-rs/aya/pull/1096 due to the EUPL license making them unable to take any inspiration from this library. Which is particularly complex because parts of these library are doing sort of the standard thing, but if anyone looks at it they become unable to do a cleanroom implementation.

Personally, I think having this support in aya would be great for the Rust community at large since it results in a single library where people can do ebpf-related things, and aya has been on the cutting edge of making cool ebpf functionality in Rust for quite a while now with an increasingly large feature set and userbase.

I don't know if the right solution is changing the license, or doing something else, but I'd be curious to hear your opinion on if there's a path forward here.

**Comment by djahandarie** [comment]
Thanks for the answer. (To be clear I'm not someone with a line of business built on top of any of this code either, so I'm not sure if your comment was meant to be directed to me or just more generally written...?)

**Comment by OliverGavin** [comment]
@djahandarie I have had success with the xdp crate in the s2n project. There is an AF_XDP socket there. Not as pretty but I was talking to the main dev on that project and he is open to making it a general purpose library for AF_XDP sockets so when I get time I will make an alternative to this based on s2n which I hope can be used in Aya without any license constraints.

**Comment by HeroicKatora** [comment]
@OliverGavin I find two glaring soundness problems minutes after looking. I don't see that library successfully trying to be safe by construction or applying a formal reasoning style to their safety analysis. For the sake of the internet, please don't let it touch TLS / safety relevant code bases in this current state. And review carefully if it fundamentally changed to be able to deliver such in future versions.

**Comment by HeroicKatora** [comment]
> Anything in particular you want to call out so it can be made safer for everyone? 

Yes, I've opened issues for everything I found. It does not seem inherently flawed but some interfaces are not as safe as they are purported to be, there are ways to UB using just safe functions. On first sight `xdp` is definitely not polished. Several `SAFETY` preconditions fail to be exhaustive enough for their actual requirements (not reported this yet as it may be a little sophistry rather than helpful, I'd rather make a PR). For instance getting a [packet by descriptor](https://docs.rs/xdp/latest/xdp/umem/struct.Umem.html#method.packet) requires one to be truthful about the options in `xdp_desc`, otherwise adjusting tx_timestamps in the resulting packet is out-of-bounds, but the method only lists the generic lifetime requirement .

From an idiomatic point there's also some minor issues that complicate analysis unecessarily. It uses the `core::index::Index` trait to expose an _unsafe_ array indexing operation on the rings. That is unorthodox to say the least. They gain a little syntactic sugar (`self.ring[i & mask]` vs `self.ring.buf(i & mask)`) but no longer can determine that said method _does_ require a safety invariants. Trait impls can also 'escape' into the public API without rustc complaints whereas dedicated methods have one very obvious (crate-)privacy definition. It doesn't currently surface _as far as I can tell_ but it's not idiomatic and avoidable. Secondly, there seems to be no comment regarding their use of atomic memory order.

I can't really _say_ whether it is broken, or not. That itself is a problematic thing to me. Currently the API for packets requires at least some  `unsafe` function calls so it's really hard to even find any interface that _could_ have concrete immediate problems.

**Comment by OliverGavin** [comment]
Oh, I was referring to this crate actually... https://crates.io/crates/s2n-quic-xdp


Totally does not have an ergonomic interface yet, but also don't think that is intended.
